### PR TITLE
8298610: Refactor archiving of ConstantPool::resolved_references()

### PR DESCRIPTION
**Pull request by iklam**
This is another prerequisite for [JDK-8296344](https://bugs.openjdk.org/browse/JDK-8296344).

Before this PR, when archiving the objArray of `ConstantPool::resolved_references()` 
- We call `HeapShared::find_archived_heap_object()` to look up archived copies of the Strings.
- We allocate a copy of the objArray inside a safepoint.

This violates the requirements [1] and [2] as stated in [JDK-8298600](https://bugs.openjdk.org/browse/JDK-8298600). 

After this PR, we no longer do the above.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298610](https://bugs.openjdk.org/browse/JDK-8298610): Refactor archiving of ConstantPool::resolved_references()

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [67fee222](https://git.openjdk.org/jdk/pull/12021/files/67fee2224115aba584c7cddf3dac60d8c38c7aae)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to [67fee222](https://git.openjdk.org/jdk/pull/12021/files/67fee2224115aba584c7cddf3dac60d8c38c7aae)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12021/head:pull/12021` \
`$ git checkout pull/12021`

Update a local copy of the PR: \
`$ git checkout pull/12021` \
`$ git pull https://git.openjdk.org/jdk pull/12021/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12021`

View PR using the GUI difftool: \
`$ git pr show -t 12021`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12021.diff">https://git.openjdk.org/jdk/pull/12021.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/cds/heapShared.cpp [hunk 1]
```diff
@@ -289,6 +290,15 @@ void HeapShared::clear_root(int index) {
   }
 }
 
+bool HeapShared::is_too_large_to_archive(oop o) {
+  // TODO: To make the CDS heap mappable for all collectors, this function should
+  // reject objects that may be too large for *any* collector.
+  assert(UseG1GC, "implementation limitation");
+  size_t sz = align_up(o->size() * HeapWordSize, ObjectAlignmentInBytes);
+  size_t max = /*G1*/HeapRegion::min_region_size_in_words() * HeapWordSize;
```
Comment (dholmes-ora) [inline src/hotspot/share/cds/heapShared.cpp:298]: This code needs to be inside an `INCLUDE_G1GC` guard does it not?

Comment (iklam) [reply src/hotspot/share/cds/heapShared.cpp:298]: This code is inside `#if INCLUDE_CDS_JAVA_HEAP`, which is enabled only if `INCLUDE_G1GC` is defined.

We have other code within `INCLUDE_CDS_JAVA_HEAP` that unconditionally access G1 APIs. These will be fixed when we remove all dependencies of G1 from the CDS heap dumping code (see [JDK-8296344](https://bugs.openjdk.org/browse/JDK-8296344) )

Comment (dholmes-ora) [reply src/hotspot/share/cds/heapShared.cpp:298]: I see - thanks.

**Review by dholmes-ora — APPROVED** [review]
This seems fine. Thanks.

**Review by calvinccheung — APPROVED** [review]
LGTM

**Comment by iklam** [comment]
Thanks @dholmes-ora and @calvinccheung for the review.
/integrate

**Pull request closed without merging**